freeipa/src/main/java/com/sequenceiq/freeipa/service/user/UserService.java

```diff
@@ -36,7 +42,50 @@
     @Inject
     private FreeIpaClientFactory freeIpaClientFactory;
 
+    @Inject
+    private ThreadBasedUserCrnProvider threadBaseUserCrnProvider;
+
+    @Inject
+    private CrnService crnService;
+
+    @Inject
+    private UmsUsersStateProvider umsUsersStateProvider;
+
+    @Inject
+    private FreeIpaUsersStateProvider freeIpaUsersStateProvider;
+
     public SynchronizeUsersResponse synchronizeUsers(SynchronizeUsersRequest request) {
+        String accountId = crnService.getCurrentAccountId();
+        List<Stack> stacks = stackService.getAllByAccountId(accountId);
+        LOGGER.debug("Found {} stacks for account {}", stacks.size(), accountId);
+        Set<String> environmentsFilter = request.getEnvironments();
+        if (!environmentsFilter.isEmpty()) {
+            stacks = stacks.stream()
+                    .filter(stack -> environmentsFilter.contains(stack.getEnvironmentCrn()))
+                    .collect(Collectors.toList());
+        }
+        for (Stack stack : stacks) {
```

Comment:
Doing this one environment at a time entails duplicating a fair amount of potentially expensive interaction with UMS: we iterate over all the users in the account for each environment, and fetch rights (and groups) for each user repeatedly for each environment.

Can we compute the UMS state for all environments in one pass, then for each environment diff the IPA state (which is unique) and sync?

Reply:
captured in DISTX-186